### DISCUSSION TITLE
Question on Test pipelines

### DISCUSSION THREAD

**Discussion by ShuffleWire** [cyberbotics/webots#5892] [General]
Hello !

I've not so much experience with CI/CD pipeline / Docker, made only a bit at work, but I've a question :    
Seen for instance https://github.com/cyberbotics/webots/actions/runs/4173526099/jobs/7225963160
At the start, I can read that some packages are installed, or even downloaded each time (for instance "django" python package, which is the biggest one)

First thing first : Am I correct ? Does the packages get pulled each time ?
If so, could it slow down significantly pipeline run, by having to download/rebuild the whole thing each time ?
Could those download also be any sort of cost ?
Do you target specific version of the pulled packages ? Otherwise could an update on their end impact Webots ?

That is just an uneducated guess and question :)

Thanks !

**Comment by omichel** [comment]
We are running the CI tests in the GitHub machines where the libraries we need are not installed by default in the image and we cannot provide our own images as far as I know. Even if we could, we would have to maintain them which is a significant overhead.

Therefore these installation steps are mandatory. The tests are running on GitHub machines and thus the downloads are likely cached, so that they run very fast and don't incur too much costs for GitHub.

**Reply by ShuffleWire** [reply]
Ok, thank for the feedback (Your answers went under my radar, but thank for the quick feedback !)
